File: src/lib/manifest_lock.js
'use strict';

/**
 * manifest_lock.js — cross-language file-locking + atomic-write helper for
 * src/strategies/manifest.json (and any other JSON file that gets written
 * by both Node and Python processes).
 *
 * Why this exists:
 *   manifest.json was getting corrupted in two distinct ways:
 *     1. Concurrent writers (lifecycle.py auto_backtest + saturday_brain.js
 *        _stage running in the same wall-clock second) producing interleaved
 *        bytes — the Python writer's `write_text()` was non-atomic, and even
 *        the JS writers' read-modify-write cycle suffered the lost-update
 *        problem when racing against Python writes.
 *     2. JSON nesting bugs from old code paths that did string-concat instead
 *        of object-merge — those are fixed but the corruption pattern looked
 *        the same as a torn write, so they were misdiagnosed.
 *
 * The fix has two parts (both required, both implemented here):
 *
 *   PART 1 — File-level mutex via O_EXCL lockfile.
 *     Both Node and Python use the SAME lockfile path and the SAME O_EXCL
 *     atomic-create semantics. Whichever process creates the lockfile first
 *     wins; others retry with backoff. Stale locks (holder process died) are
 *     detected via PID liveness check + a generous 60s lockfile mtime cutoff.
 *
 *   PART 2 — Atomic write via tmp + rename.
 *     Even with the lock held, the write itself uses `O_WRONLY | O_CREAT |
 *     O_TRUNC` on a sibling `.tmp` file, then `rename()` — which on POSIX
 *     filesystems is atomic. A crashed writer leaves a `.tmp` orphan but
 *     never a half-written manifest.json.
 *
 * Lock contract (between Node and Python):
 *   - Lockfile path = `<target>.lock`
 *   - Lockfile content = `<PID>\n<ISO timestamp>\n<actor tag>` (3 lines)
 *   - Acquire: `fs.openSync(lockfile, O_WRONLY | O_CREAT | O_EXCL)`
 *   - Release: `fs.unlinkSync(lockfile)`
 *   - Stale: lockfile mtime > 60s OR `kill -0 <pid>` returns ESRCH → force-unlink and retry
 *
 * Public API:
 *   await withManifestLock(targetPath, async (currentManifest) => {
 *     currentManifest.strategies['foo'] = bar;
 *     return currentManifest;        // helper writes back atomically
 *   }, { actor: 'saturday_brain' });
 *
 *   // Or use the lower-level acquire/release for fine-grained control:
 *   const release = await acquireLock(targetPath, { actor: '...' });
 *   try { ... } finally { release(); }
 */

const fs   = require('fs');
const path = require('path');

const LOCK_TIMEOUT_MS    = 60_000;   // stale-lock cutoff
const ACQUIRE_TIMEOUT_MS = 30_000;   // give up acquiring after 30s
const POLL_BASE_MS       = 25;       // initial poll interval
const POLL_MAX_MS        = 500;      // cap on backoff

function _lockPath(targetPath) {
  return targetPath + '.lock';
}

function _isProcessAlive(pid) {
  if (!pid || !Number.isFinite(pid) || pid <= 1) return false;
  try {
    // signal 0 = check existence + permissions, doesn't actually signal
    process.kill(pid, 0);
    return true;
  } catch (e) {
    if (e.code === 'EPERM') return true;   // exists, just not ours to signal
    return false;                          // ESRCH or anything else → dead
  }
}

function _readLockMeta(lockfile) {
  try {
    const stat = fs.statSync(lockfile);
    const txt  = fs.readFileSync(lockfile, 'utf8');
    const [pidStr, ts, actor] = txt.split('\n');
    return { pid: parseInt(pidStr, 10), ts, actor: actor || '?', mtimeMs: stat.mtimeMs };
  } catch (_) { return null; }
}

function _maybeClearStaleLock(lockfile) {
  const meta = _readLockMeta(lockfile);
  if (!meta) return false;     // lockfile vanished
  const ageMs = Date.now() - meta.mtimeMs;
  const stale = ageMs > LOCK_TIMEOUT_MS || !_isProcessAlive(meta.pid);
  if (!stale) return false;
  try {
    fs.unlinkSync(lockfile);
    process.stderr.write(
      `[manifest_lock] cleared stale lock at ${lockfile} ` +
      `(age=${Math.round(ageMs / 1000)}s, pid=${meta.pid}/${meta.actor}, ` +
      `alive=${_isProcessAlive(meta.pid)})\n`
    );
    return true;
  } catch (_) { return false; }
}

function _sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

/**
 * Acquire the lock. Returns a release function. Throws on timeout.
 *
 * @param {string} targetPath  the file being protected (lockfile is
 *                             targetPath + '.lock')
 * @param {object} opts
 *   - actor: string identifying who's holding (logs only, default 'unknown')
 *   - timeoutMs: max wait for acquire (default 30000)
 */
async function acquireLock(targetPath, opts = {}) {
  const lockfile = _lockPath(targetPath);
  const actor    = opts.actor    || 'unknown';
  const deadline = Date.now() + (opts.timeoutMs || ACQUIRE_TIMEOUT_MS);
  let backoff = POLL_BASE_MS;

  // Ensure target dir exists (lockfile dir == target dir).
  try { fs.mkdirSync(path.dirname(lockfile), { recursive: true }); } catch (_) {}

  while (true) {
    let fd;
    try {
      fd = fs.openSync(lockfile, fs.constants.O_WRONLY | fs.constants.O_CREAT | fs.constants.O_EXCL, 0o600);
      const payload = `${process.pid}\n${new Date().toISOString()}\n${actor}\n`;
      fs.writeSync(fd, payload);
      fs.closeSync(fd);
      // Success.
      return function release() {
        try { fs.unlinkSync(lockfile); } catch (_) {}
      };
    } catch (e) {
      if (e.code !== 'EEXIST') throw e;
      // Held by someone else — check if stale.
      _maybeClearStaleLock(lockfile);
      if (Date.now() > deadline) {
        const meta = _readLockMeta(lockfile);
        const heldBy = meta ? `pid=${meta.pid} actor=${meta.actor} ts=${meta.ts}` : 'unknown';
        throw new Error(`manifest_lock acquire timed out after ${opts.timeoutMs || ACQUIRE_TIMEOUT_MS}ms; held by ${heldBy}`);
      }
      await _sleep(backoff);
      backoff = Math.min(backoff * 2, POLL_MAX_MS);
    }
  }
}

/**
 * High-level wrapper. Reads the manifest under lock, passes it to the
 * callback, writes the (possibly mutated) result back atomically. The
 * callback may return a NEW manifest object or mutate the passed one in
 * place (and return undefined / the same reference).
 *
 * @param {string}   targetPath
 * @param {function} cb         async (manifest) => manifest|undefined
 * @param {object}   opts       { actor, timeoutMs, encoding ('utf8') }
 * @returns whatever the callback returned (after the write completes)
 */
async function withManifestLock(targetPath, cb, opts = {}) {
  const release = await acquireLock(targetPath, opts);
  try {
    const encoding = opts.encoding || 'utf8';
    let raw;
    try {
      raw = fs.readFileSync(targetPath, encoding);
    } catch (e) {
      if (e.code === 'ENOENT') raw = '{}';   // first-write
      else throw e;
    }
    let manifest;
    try {
      manifest = JSON.parse(raw);
    } catch (e) {
      throw new Error(`manifest_lock: target ${targetPath} is not valid JSON — ${e.message}`);
    }
    const result = await cb(manifest);
    const toWrite = (result === undefined) ? manifest : result;

    const tmp = targetPath + '.tmp';
    fs.writeFileSync(tmp, JSON.stringify(toWrite, null, 2), { encoding });
    fs.renameSync(tmp, targetPath);
    return toWrite;
  } finally {
    release();
  }
}

/**
 * Synchronous read under lock. Useful for callers that just want a coherent
 * snapshot without the read-modify-write cycle (e.g. server.js fetching the
 * manifest for an /api/strategies response — though that endpoint can also
 * just do a plain readFileSync since concurrent writes are atomic).
 */
async function readManifestUnderLock(targetPath, opts = {}) {
  const release = await acquireLock(targetPath, opts);
  try {
    return JSON.parse(fs.readFileSync(targetPath, opts.encoding || 'utf8'));
  } finally {
    release();
  }
}

module.exports = {
  acquireLock,
  withManifestLock,
  readManifestUnderLock,
  _internals: { _isProcessAlive, _readLockMeta, _maybeClearStaleLock, _lockPath },
};

File: src/strategies/_manifest_lock.py
"""
_manifest_lock.py — Python counterpart to src/lib/manifest_lock.js.

Both implementations MUST agree on:
  - Lockfile path:    `<target>.lock`
  - Lockfile content: 3 lines = "<PID>\\n<ISO timestamp>\\n<actor>\\n"
  - Acquire mode:     atomic O_EXCL creation
  - Stale detection:  mtime > 60s OR PID not alive
  - Atomic write:     write `<target>.tmp` then `os.replace(tmp, target)`

If you change the contract here, update manifest_lock.js too.
"""

from __future__ import annotations

import contextlib
import errno
import json
import os
import signal
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable, Iterator, Optional

LOCK_TIMEOUT_S    = 60.0    # stale-lock cutoff
ACQUIRE_TIMEOUT_S = 30.0    # give up acquiring after 30s
POLL_BASE_S       = 0.025
POLL_MAX_S        = 0.5


def _lock_path(target_path: str | Path) -> str:
    return str(target_path) + ".lock"


def _is_process_alive(pid: int) -> bool:
    if not pid or pid <= 1:
        return False
    try:
        os.kill(pid, 0)   # signal 0 = check existence
        return True
    except ProcessLookupError:
        return False
    except PermissionError:
        return True       # exists, just not ours
    except OSError:
        return False


def _read_lock_meta(lockfile: str) -> Optional[dict]:
    try:
        st = os.stat(lockfile)
        with open(lockfile, "r", encoding="utf-8") as f:
            txt = f.read()
        parts = txt.split("\n")
        pid_str = parts[0] if parts else ""
        ts      = parts[1] if len(parts) > 1 else ""
        actor   = parts[2] if len(parts) > 2 else "?"
        return {
            "pid":      int(pid_str) if pid_str.isdigit() else 0,
            "ts":       ts,
            "actor":    actor or "?",
            "mtime_s":  st.st_mtime,
        }
    except FileNotFoundError:
        return None
    except (OSError, ValueError):
        return None


def _maybe_clear_stale_lock(lockfile: str) -> bool:
    meta = _read_lock_meta(lockfile)
    if not meta:
        return False
    age_s = time.time() - meta["mtime_s"]
    stale = age_s > LOCK_TIMEOUT_S or not _is_process_alive(meta["pid"])
    if not stale:
        return False
    try:
        os.unlink(lockfile)
        import sys
        sys.stderr.write(
            f"[_manifest_lock] cleared stale lock at {lockfile} "
            f"(age={age_s:.0f}s, pid={meta['pid']}/{meta['actor']}, "
            f"alive={_is_process_alive(meta['pid'])})\n"
        )
        return True
    except OSError:
        return False


def _acquire_lock(target_path: str | Path, *, actor: str = "unknown",
                  timeout_s: float = ACQUIRE_TIMEOUT_S) -> Callable[[], None]:
    """Acquire the lock; returns a release callable. Raises TimeoutError on
    timeout."""
    lockfile = _lock_path(target_path)
    Path(lockfile).parent.mkdir(parents=True, exist_ok=True)
    deadline = time.time() + timeout_s
    backoff  = POLL_BASE_S

    while True:
        try:
            fd = os.open(lockfile, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
            try:
                ts = datetime.now(timezone.utc).isoformat()
                payload = f"{os.getpid()}\n{ts}\n{actor}\n".encode()
                os.write(fd, payload)
            finally:
                os.close(fd)

            def release() -> None:
                try:
                    os.unlink(lockfile)
                except OSError:
                    pass
            return release

        except FileExistsError:
            _maybe_clear_stale_lock(lockfile)
            if time.time() > deadline:
                meta = _read_lock_meta(lockfile)
                held_by = (
                    f"pid={meta['pid']} actor={meta['actor']} ts={meta['ts']}"
                    if meta else "unknown"
                )
                raise TimeoutError(
                    f"manifest_lock acquire timed out after {timeout_s}s; "
                    f"held by {held_by}"
                )
            time.sleep(backoff)
            backoff = min(backoff * 2, POLL_MAX_S)


@contextlib.contextmanager
def manifest_lock(target_path: str | Path, *, actor: str = "unknown",
                  timeout_s: float = ACQUIRE_TIMEOUT_S) -> Iterator[None]:
    """Context manager. Acquires the lock; releases on exit (success or
    exception). Use this around the entire read-modify-write cycle."""
    release = _acquire_lock(target_path, actor=actor, timeout_s=timeout_s)
    try:
        yield
    finally:
        release()


def write_atomic(target_path: str | Path, payload: Any, *, indent: int = 2,
                 encoding: str = "utf-8") -> None:
    """Atomic JSON write: write to <target>.tmp then rename. Caller MUST hold
    the manifest_lock when calling this; otherwise the read-modify-write
    cycle is racy even though the write itself is atomic."""
    target = Path(target_path)
    tmp = Path(str(target) + ".tmp")
    tmp.write_text(json.dumps(payload, indent=indent), encoding=encoding)
    os.replace(tmp, target)


def read_under_lock(target_path: str | Path, *, actor: str = "reader",
                    timeout_s: float = ACQUIRE_TIMEOUT_S) -> dict:
    """Read + parse JSON under lock. Useful when a caller needs a coherent
    snapshot without write intent."""
    with manifest_lock(target_path, actor=actor, timeout_s=timeout_s):
        with open(target_path, "r", encoding="utf-8") as f:
            return json.load(f)


def with_manifest_lock(target_path: str | Path, mutator: Callable[[dict], Optional[dict]],
                       *, actor: str = "unknown", timeout_s: float = ACQUIRE_TIMEOUT_S) -> dict:
    """High-level read-modify-write convenience. Equivalent to:

        with manifest_lock(target_path):
            data = json.load(open(target_path))
            result = mutator(data) or data
            write_atomic(target_path, result)

    Returns the written payload."""
    with manifest_lock(target_path, actor=actor, timeout_s=timeout_s):
        try:
            with open(target_path, "r", encoding="utf-8") as f:
                data = json.load(f)
        except FileNotFoundError:
            data = {}
        result = mutator(data)
        if result is None:
            result = data
        write_atomic(target_path, result)
        return result
